components/async-container/async-container-mixin.js

```diff
@@ -0,0 +1,78 @@
+
+export const asyncStates = {
+	initial: 'initial',
+	pending: 'pending',
+	success: 'success',
+	failure: 'failure'
+};
+
+export const AsyncContainerMixin = superclass => class extends superclass {
+
+	static get properties() {
+		return {
+			asyncState: { type: String }
+		};
+	}
+
+	constructor() {
+		super();
+		this._initializeAsyncState();
+		if (!this.asyncContainer) {
+			this.addEventListener('pending-state', this._handleAsyncItemState.bind(this));
+		}
+	}
+
+	firstUpdated() {
+		if (!this.asyncContainer) return;
+		const container = this.asyncContainer();
+		if (container) {
+			container.addEventListener('pending-state', this._handleAsyncItemState.bind(this));
+		}
+	}
+
+	resetAsyncState() {
+		this._initializeAsyncState();
+	}
+
+	async _handleAsyncItemState(e) {
+		const promise = e.detail.promise;
+
+		if (!promise) return;
+		if (this.asyncState === asyncStates.success
+			|| this.asyncState === asyncStates.failure) return;
+
+		this._asyncPromises.push(promise);
+		this._asyncCounts.pending++;
+		this.asyncState = asyncStates.pending;
+
+		try {
+			await promise;
+			if (this._asyncPromises.indexOf(promise) !== -1) this._asyncCounts.fulfilled++;
```

Comment:
The purpose these `(this._asyncPromises.indexOf(promise) !== -1)` checks is to ensure the promise is still being tracked before updating the counts, because it is possible that the state is reset (via `resetAsyncState()`) while promises are pending. In such case, when those promises reach their final state, they are effectively ignored. At first I consider tracking the pending, fulfilled, and rejected promises, but I did not want to recreate arrays every time a promise is resolved or rejected, since there could be many async items being tracked.